XenAdmin/Commands/InstallToolsCommand.cs

```diff
@@ -298,54 +297,32 @@ public bool ConfirmAndExecute()
 
         public static bool CanExecute(VM vm)
         {
-            var virtualisationStatus = vm.GetVirtualisationStatus();
-
-            return vm != null && !vm.is_a_template && !vm.Locked &&
-                !virtualisationStatus.HasFlag(VM.VirtualisationStatus.UNKNOWN) &&
-                (!virtualisationStatus.HasFlag(VM.VirtualisationStatus.IO_DRIVERS_INSTALLED) || !virtualisationStatus.HasFlag(VM.VirtualisationStatus.MANAGEMENT_INSTALLED)) &&
-                vm.power_state == vm_power_state.Running && !ResidentHostIsOlderThanMaster(vm)
-                && CanViewVMConsole(vm.Connection);
-        }
+            if (vm == null || vm.is_a_template || vm.Locked || vm.power_state != vm_power_state.Running)
+                return false;
 
-        private static bool ResidentHostIsOlderThanMaster(VM vm)
-        {
-            var vmHome = vm.Home();
-            return vmHome != null && Helpers.IsOlderThanMaster(vmHome);
-        }
+            var vStatus = vm.GetVirtualisationStatus();
 
-        public static bool CanExecuteAll(List<VM> vms)
-        {
-            foreach (VM vm in vms)
-            {
-                if (!CanExecute(vm))
-                    return false;
-            }
-            return true;
-        }
-
-        private static bool CanViewVMConsole(XenAdmin.Network.IXenConnection xenConnection)
-        {
-            if (xenConnection.Session == null)
+            if (vStatus.HasFlag(VM.VirtualisationStatus.UNKNOWN) ||
+                vStatus.HasFlag(VM.VirtualisationStatus.IO_DRIVERS_INSTALLED) && vStatus.HasFlag(VM.VirtualisationStatus.MANAGEMENT_INSTALLED))
                 return false;
 
-            RbacMethodList r = new RbacMethodList("http/connect_console");
-            if (Role.CanPerform(r, xenConnection, false))
-                return true;
+            var vmHome = vm.Home();
+            if (vmHome != null && Helpers.IsOlderThanMaster(vmHome))
+                return false;
 
-            return false;
+            //whether RBAC allows connection to the VM's console
+            return vm.Connection.Session != null &&
+                   Role.CanPerform(new RbacMethodList("http/connect_console"), vm.Connection, out _, false);
         }
 
         protected override bool CanExecuteCore(SelectedItemCollection selection)
         {
-            return selection.AllItemsAre<VM>() && selection.AtLeastOneXenObjectCan<VM>(CanExecute);
+            return selection.Count > 0 &&
+                   selection.All(v => v.XenObject is VM vm &&

```

Comment:
I changed the "at least one" to "all" because for cross pool selection of pre- and post- stockholm pools the experience will be confusing for the user.